pootle/apps/pootle_app/management/commands/create_project.py

```diff
@@ -0,0 +1,171 @@
+            dest="sourcelang",
+            default="en",
+            help=("Source language. Examples: [en | es | fr | ...]."
+                  "Default: %(default)s"),
+        )
+        parser.add_argument(
+            "--report-email",
+            action="store",
+            default="",
+            dest="contact",
+            help="Contact email for reports. Example: admin@mail.com.",
+        )
+        parser.add_argument(
+            "--disabled",
+            action="store_true",
+            dest="disabled",
+            help="Does the project start disabled?",
+        )
+        parser.add_argument(
+            "--checkstyle",
+            action="store",
+            dest="checkstyle",
+            choices=PROJECT_CHECKERS.keys(),
+            default="standard",
+            help="Quality check styles. Example: {0}. Default: %(default)s"
+                 .format(" | ".join(PROJECT_CHECKERS.keys())),
+        )
+
+    @transaction.atomic
+    def handle(self, **options):
+        """Immitates the behaviour of the admin interface
+        for creating a project from command line.
+        """
+        self.check_project_options(**options)
+        self.set_project_config(
+            self.create_project(**options), **options)
+
+    def check_project_options(self, **options):
+        if options["contact"]:
+            try:
+                validate_email(options["contact"])
+            except ValidationError as e:
+                if options["traceback"]:
+                    traceback.print_exc()
+                raise CommandError(e)
+        if options["fs_type"] != "localfs" and options["fs_url"] == "":
+            raise CommandError("Parameter --fs-url is mandatory "
+                               "when --fs-type is not `localfs`")
+
+    def create_project(self, **options):
+        try:
+            return Project.objects.create(
+                code=options["code"],
+                fullname=options["name"] or options["code"].capitalize(),
+                checkstyle=options["checkstyle"],
+                source_language=self.get_source_language(**options),
+                filetypes=options["filetypes"] or ["po"],
+                report_email=options["contact"],
+                disabled=options["disabled"])
+        except ValidationError as e:
+            if options["traceback"]:
+                traceback.print_exc()
+            raise CommandError(e)
+
+    def get_source_language(self, **options):
+        try:
+            return Language.objects.get(code=options["sourcelang"])
+        except Language.DoesNotExist:
+            raise CommandError("Source language %s does not exist"
+                               % options["sourcelang"])
+
+    def set_project_config(self, project, **options):
+        mapping = (
+            dict(default=FS_PRESETS[options["preset_mapping"]][0])
+            if options["preset_mapping"]
+            else dict(default=options["mapping"]))
```

Comment:
Is the custom mapping validated in any way?

Reply:
yes, when its set

Reply:
nvm, just saw the test.